### DISCUSSION TITLE
How to run tests after building MOOSE with singularity?

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#26101] [Q&A Getting Started]
Hi all, I'm trying to build MOOSE with singularity on the cluster, it builds successfully by pulling out docker image:

`singularity build moose.sif docker://idaholab/moose:latest`

Then I would like to run tests to verify the build, I try to search a similar command as given in MOOSE docker install page:

`docker run -ti idaholab/moose:latest /bin/bash -c 'cd test; ./run_tests'`

But I'm unable to find a similar command to get access to the executable. I wonder if anyone has experience? Thanks!

**Comment by loganharbour** [comment]
@chunhuizhao478 we are in the slow process of phasing out our docker images in lieu of our apptainer images, and the documentation has not been updated accordingly.

If you run:

```
singularity build moose.sif docker://idaholab/moose-dev:latest
```

you'll build the latest moose environment (this is a WIP, but is nearly done). You can then enter that environment, clone, and then build moose.

What is your use case here? I'm glad to help if you provide more context.

**Reply by chunhuizhao478** [reply]
Thanks @loganharbour. We would like to build singularity container of MOOSE and combine my own application on HPC cluter (Expanse in UCSD) for a project (Quakeworx). Since I'm new to singularity, it's quite confusing at this stage. The ultimate goal is to have MOOSE (with tensormechanics modulus) + my own application as a singularity container that the example code can be executed on Expanse.

So far I'm able to build moose.sif (I will try the dev you mentioned soon) on the cluster, now I would like to 
(1) test the build (use ./run_tests)
(2) combine my own application (available on GitHub)
(3) run an example code on the cluster

I'm really appreciating if you could highlight the necessary commands for achieving this. Thanks!

**Reply by loganharbour** [reply]
I'm sorry for the delay on this; the answer to your questions depends quite a bit on your HPC setup.

The most important question is that are you tying to run MOOSE across multiple nodes on your cluster? If so, the configuration for using the container is a bit more detailed.

**Comment by chunhuizhao478** [comment]
It builds successfully with `moose-dev:latest`: 

```
[czhao1@exp-1-21 ~]$ singularity build moose_dev.sif docker://idaholab/moose-dev:latest
WARNING: 'nodev' mount option set on /tmp, it could be a source of failure during build process
INFO:    Starting build...
Getting image source signatures
Copying blob 02038a90d4b6 skipped: already exists  
Copying config 3beeea92be done  
Writing manifest to image destination
Storing signatures
2023/11/20 07:20:45  info unpack layer: sha256:02038a90d4b68c7c5a4918742e510bdf1aa554dd1a6e8de29f9c3719d6a97c11
INFO:    Creating SIF file...
INFO:    Build complete: moose_dev.sif
```

If I do `singularity shell moose_dev.sif; ls \`:

```
[moose_dev][~]> ls /
bin   dev	  environment  home  lib64	 media	none  proc  run			       sbin	    srv  tmp  var
boot  Dockerfile  etc	       lib   lost+found  mnt	opt   root  run_singularity2docker.sh  singularity  sys  usr
```

But there is no moose, so the next step is to build moose inside, how should i achieve it?
